### DISCUSSION TITLE
How about custom node view render mode?

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-glsp/glsp#929] [Q&A]
Hi,
    I create a custom node view for my custom node in glsp. I have two questions
question one is about the key of VNode
my custom node view as follow 
![image](https://user-images.githubusercontent.com/4537809/221127903-9dfc00f2-5e87-437e-a7ac-6efb1be4fea4.png)
if I do not set key of vnode, my custom node view can update value normally
![2023-02-24_16-32-07](https://user-images.githubusercontent.com/4537809/221133473-49c2e974-3e98-42fb-bf69-104b872d27c8.gif)

if I set key of vnode and put mouse over my custom node, It will always stay in initial state
![2023-02-24_16-00-34](https://user-images.githubusercontent.com/4537809/221129940-22bde63a-9f70-4ad0-8635-3a959b6f0bee.gif)

My code as follow 
![image](https://user-images.githubusercontent.com/4537809/221130119-303d27bd-8804-4807-9e5f-0f03aa74ddc8.png)
![image](https://user-images.githubusercontent.com/4537809/221130142-ff81f7ea-f9a4-47dc-bca1-637bb5e513c8.png)

Question two is about the injection mode of custom node view. 
Does it use singleton inject for custom node view? I create a collection to get information in my gaugeNodeView. I create two guageNodeView instance in frontend. finally I get all information of two instances in the same collection

one strange thing is that when I open toggle develop tool and find two diagram containers, I just create one node view instance 
int frontend panel.
![image](https://user-images.githubusercontent.com/4537809/221131366-fadd3fbb-e89a-4e36-a6b0-39a39a01a16e.png)
![image](https://user-images.githubusercontent.com/4537809/221131323-83f4e1ad-4a86-44a5-8789-7ba9adafdece.png)

**Comment by planger** [comment]
Hi,

regarding your second question:
Sprotty renders the diagram twice, once in a hidden `div` to obtain the sizes of all elements and once in the `div` that is visible to the user. This is needed for layouters, since they need to know the sizes of all elements, before they can position elements based on the size, e.g. in a horizontal box or in a vertical box. Without actually "rendering" them, the sizes of elements cannot be determined properly, as they may depend on CSS properties, such as font size, etc.

regarding your first question:
Unfortunately, I'm not sure... just from looking a the code, I didn't spot anything suspicious without further debugging into it. Also it seems like a snabbdom-related issue, so I'd probably ask that question there. https://github.com/snabbdom/snabbdom

**Reply by wss29** [reply]
Hi,
    Thanks greatly for your reply
   I debug and find that the gauge node view constructor is invoked only once at the application start, after that I add more instances in frontend the constructor never be invoked.  It seems that it uses singleton inject mode, could it be changed to transient?  singleton is not suite for my situation, because I have a different state for different node view.

**Reply by planger** [reply]
Oh, right, the views are singletons, as they are supposed to be stateless. I don't think this can be changed easily.
